Refactoring of IncrementalBuilder

majocha opened the pull request:
I refactored `BoundModel` by removing `TcInfoState`. Now there is no `PartialState` / `FullState`, we instead create separate `GraphNodes` computing `TcInfo` and `TcInfoExtras`. When computing `TcInfo` this always skips implementations whenever possible. Computing `TcInfoExtras` OTOH always does full type check of the file, so previously skipped implementations can be checked in parallel.

I made it more explicit when the skipping of implementation during parsing and type check happens. Separate functions return parsing result placeholder and partial type check info for skipped implementations.

Build graph is now updated in a single pass `mapFold` in `computeStampedFileNames`. I find this easier to comprehend, but that's individual matter probably.

A `Slot` record is introduced to hold together parsing and typechecking GraphNodes for particular file slots.

Removed parsing cache because it's now irrelevant. Parsing step has it's own GraphNode which ensures we never reparse unless the client or the file system notifies there was source change.

`enablePartialTypeChecking` setting is also now mostly irrelevant, as TcInfoExtras are computed on demand. Although it's easy to change this behavior to parallelized eager full type check when `enablePartialTypeChecking = false`. I'm not sure what would be best here.


This performs somewhat better in cases when there are backing signatures. There is still room for performance gain if we do the parsing eagerly instead of on-demand, and even more if we could use build graph from #14494.

Some benchmarks:

main:

```
|                       Method | FastFindReferences | EmptyCache |         Mean |      Error |     StdDev |       Gen0 |      Gen1 |   Allocated |
|----------------------------- |------------------- |----------- |-------------:|-----------:|-----------:|-----------:|----------:|------------:|
| FindAllReferences_MediumCase |              False |      False |     93.77 ms |   1.428 ms |   1.266 ms |          - |         - |     2.53 MB |
| FindAllReferences_MediumCase |              False |       True | 10,664.60 ms | 168.664 ms | 149.516 ms | 11000.0000 | 2000.0000 | 12900.45 MB |
| FindAllReferences_MediumCase |               True |      False |    238.00 ms |   4.718 ms |   5.966 ms |  1000.0000 |         - |   173.17 MB |
| FindAllReferences_MediumCase |               True |       True |  5,518.70 ms |  47.706 ms |  42.291 ms |  9000.0000 | 2000.0000 |  3557.88 MB |
```

this PR:

```
|                       Method | FastFindReferences | EmptyCache |         Mean |      Error |     StdDev |       Median |      Gen0 |      Gen1 |     Allocated |
|----------------------------- |------------------- |----------- |-------------:|-----------:|-----------:|-------------:|----------:|----------:|--------------:|
| FindAllReferences_MediumCase |              False |      False |     89.89 ms |   0.671 ms |   0.594 ms |     90.07 ms |         - |         - |    2942.72 KB |
| FindAllReferences_MediumCase |              False |       True | 10,705.55 ms | 209.875 ms | 294.216 ms | 10,532.61 ms | 6000.0000 | 2000.0000 | 7980685.08 KB |
| FindAllReferences_MediumCase |               True |      False |     17.20 ms |   0.370 ms |   1.018 ms |     16.94 ms |         - |         - |     856.41 KB |
| FindAllReferences_MediumCase |               True |       True |  5,297.71 ms |  37.674 ms |  35.240 ms |  5,300.07 ms | 3000.0000 | 1000.0000 | 3215518.17 KB |
```

There is some win with FastFindReferences, it seems.

more benchmarks in comments.

majocha commented:
This seems to give approx 10% gain in `ParseProjectWithChangingMiddleFile` benchmark compared to main.

majocha commented:
I'll continue with this, but first I need some more tests, see #14921, because I can completely break partial type checking and this is still green.

majocha commented:
This also needs debouncing when used with live buffers, or it will spam lots of non-cancellable parse tasks  when typing. I'll do it in another PR for VS.

majocha commented:
That failing test here was a curious case. Test was for #14396, but the failure was unrelated. It failed because of errors in synthetic project, that went uncaught before.
Something to investigate.

majocha commented:
This is green and feels snappy at first glance.
But I need to benchmark this against main when I'll have the time.

nojaf commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -285,7 +303,6 @@ type BoundModel private (tcConfig: TcConfig,
                          keepAssemblyContents, keepAllBackgroundResolutions,
                          keepAllBackgroundSymbolUses,
                          enableBackgroundItemKeyStoreAndSemanticClassification,
-                         enablePartialTypeChecking,
```
I'm a bit lost by this removal.
Are you now always assuming partial type checking?

majocha commented on src/Compiler/Service/service.fs:
```diff
@@ -191,7 +191,7 @@ type BackgroundCompiler
         keepAllBackgroundSymbolUses,
         enableBackgroundItemKeyStoreAndSemanticClassification,
-        enablePartialTypeChecking,
+        _enablePartialTypeChecking,
```
Yeah, that was kind of experimental. It irked me how the whole code was split into two paths by this setting. I think I refactored it to the state where I can quickly restore this now with little effort.

majocha replied:
Restored, 🙂. Let's see if something breaks now.

majocha commented:
Restoring `enablePartialTypeChecking` throughout also resolved problem with that test for #14396.

I may be wrong, but a quick search in this repo shows suggests we have actually zero tests with `enablePartialTypeChecking = false`?
@nojaf, that's why I briefly thought to just make it a default, but of course Rider and FSAutoComplete do use it.

nojaf commented:
> I may be wrong, but a quick search in this repo suggests we have actually zero tests with enablePartialTypeChecking = false?

That doesn't really say anything. It is part of the public FCS API and should be respected as such.
Rider indeed exposes his flag. Thanks for addressing it.

majocha commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -138,11 +139,11 @@ module IncrementalBuildSyntaxTree =
-            use _holder = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parse)
-            try           
+        let getParseTask (source: FSharpSource) =
+            backgroundTask {
```
I might have misunderstood how backgroundTask works. The intent was to start that job on thread pool as soon as possible. Async.StartAsTask?

vzarytovskii replied:
If you want it on the threadpool, you probably want to `Task.Run` it.

But i'm slightly confused here too.

Do I understand the intention correctly - you want it to start (on background) as soon as it's created, so when you need an actual result, it would be hightly probable that it's already started (or even finished)?

vzarytovskii replied:
Then, yes, you probably want to have it running via `Task.Run`.

majocha replied:
thanks @vzarytovskii, it does help, with Task.Run I can see the gains in FileCascade benchmark.

vzarytovskii commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -156,39 +157,55 @@ module IncrementalBuildSyntaxTree =
                 failwith msg
 
-        let parseOrSkip sigNameOpt =
+        let getValue source = lock source <| fun () -> cache.GetValue(source, getParseTask)
```
Just a sanity check - is shadowed `source` fine to lock on in here.

majocha replied:
Should be fine, source is cache key and getParseTask does not wait.

vzarytovskii commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -138,11 +139,11 @@ module IncrementalBuildSyntaxTree =
-            use _holder = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parse)
-            try           
+        let getParseTask (source: FSharpSource) =
+            Threading.Tasks.Task.Run( fun () ->
```
One thing which bothers me slighlty, is that now, we `lock` in some places inside incremental builder, which may lead to the thread pool exhaustion, and those tasks may not be executing.

This should help it a bit: https://github.com/dotnet/fsharp/pull/14904

majocha replied:
The lock is around a particular cache key, so it should be very granular, but I'll take a look.

majocha commented:
ParseProjectWithFullCacheClear is the benchmark that sees most gains.

main:

```|                                  Method | PartialCheck | GenerateFSI |      Mean |     Error |    StdDev | Ratio | RatioSD | Rank |       Gen0 |      Gen1 |      Gen2 |    Allocated | Alloc Ratio |
|---------------------------------------- |------------- |------------ |----------:|----------:|----------:|------:|--------:|-----:|-----------:|----------:|----------:|-------------:|------------:|
|        ParseAndCheckLastFileProjectAsIs |        False |       False |  10.76 ms |  0.073 ms |  0.048 ms |  0.01 |    0.00 |    I |          - |         - |         - |    114.38 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |        False |       False |  25.82 ms |  0.271 ms |  0.180 ms |  0.04 |    0.00 |   II |   218.7500 |   31.2500 |         - |   8357.29 KB |       0.019 |
|      ParseProjectWithChangingMiddleFile |        False |       False | 283.57 ms |  9.936 ms |  5.197 ms |  0.39 |    0.01 |  III |  5000.0000 | 1000.0000 |         - | 174423.02 KB |       0.402 |
|       ParseProjectWithChangingFirstFile |        False |       False | 544.02 ms |  3.666 ms |  2.425 ms |  0.74 |    0.01 |   IV | 10000.0000 | 2000.0000 |         - |  341819.3 KB |       0.787 |
|          ParseProjectWithFullCacheClear |        False |       False | 736.28 ms | 13.963 ms |  9.236 ms |  1.00 |    0.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 434265.21 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |        False |        True |  21.69 ms |  0.262 ms |  0.174 ms |  0.02 |    0.00 |    I |          - |         - |         - |    231.77 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |        False |        True |  38.92 ms |  0.366 ms |  0.218 ms |  0.04 |    0.00 |   II |   230.7692 |         - |         - |   9583.75 KB |       0.019 |
|      ParseProjectWithChangingMiddleFile |        False |        True | 346.65 ms | 15.736 ms | 10.409 ms |  0.39 |    0.01 |  III |  5000.0000 | 1000.0000 |         - | 195964.72 KB |       0.395 |
|       ParseProjectWithChangingFirstFile |        False |        True | 658.76 ms |  6.197 ms |  4.099 ms |  0.74 |    0.00 |   IV | 11000.0000 | 2000.0000 |         - | 382261.74 KB |       0.770 |
|          ParseProjectWithFullCacheClear |        False |        True | 893.65 ms |  6.068 ms |  3.173 ms |  1.00 |    0.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 496278.27 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |       False |  10.71 ms |  0.068 ms |  0.045 ms |  0.01 |    0.00 |    I |          - |         - |         - |    111.15 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |         True |       False |  25.85 ms |  0.227 ms |  0.150 ms |  0.04 |    0.00 |   II |   218.7500 |   31.2500 |         - |   8302.21 KB |       0.019 |
|      ParseProjectWithChangingMiddleFile |         True |       False | 293.20 ms | 41.663 ms | 27.558 ms |  0.40 |    0.04 |  III |  4500.0000 | 1000.0000 |         - | 171913.12 KB |       0.399 |
|       ParseProjectWithChangingFirstFile |         True |       False | 532.70 ms |  4.834 ms |  2.877 ms |  0.73 |    0.01 |   IV | 10000.0000 | 2000.0000 |         - | 338349.57 KB |       0.785 |
|          ParseProjectWithFullCacheClear |         True |       False | 730.82 ms | 17.076 ms | 10.162 ms |  1.00 |    0.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 430901.38 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |        True |  21.65 ms |  0.307 ms |  0.203 ms |  0.07 |    0.00 |    I |          - |         - |         - |     215.6 KB |       0.001 |
| ParseProjectWithChangingPenultimateFile |         True |        True |  30.32 ms |  0.200 ms |  0.132 ms |  0.10 |    0.00 |   II |   125.0000 |         - |         - |    4376.7 KB |       0.028 |
|      ParseProjectWithChangingMiddleFile |         True |        True |  88.46 ms |  0.517 ms |  0.342 ms |  0.30 |    0.01 |  III |   833.3333 |  333.3333 |         - |  30642.21 KB |       0.194 |
|       ParseProjectWithChangingFirstFile |         True |        True | 150.05 ms |  9.406 ms |  6.222 ms |  0.50 |    0.02 |   IV |  1500.0000 |  750.0000 |         - |  53467.72 KB |       0.339 |
|          ParseProjectWithFullCacheClear |         True |        True | 299.00 ms |  9.226 ms |  5.490 ms |  1.00 |    0.00 |    V |  1000.0000 | 1000.0000 | 1000.0000 | 157904.95 KB |       1.000 |
```

this PR:

```
|                                  Method | PartialCheck | GenerateFSI |      Mean |     Error |    StdDev | Ratio | Rank |       Gen0 |      Gen1 |      Gen2 |    Allocated | Alloc Ratio |
|---------------------------------------- |------------- |------------ |----------:|----------:|----------:|------:|-----:|-----------:|----------:|----------:|-------------:|------------:|
|        ParseAndCheckLastFileProjectAsIs |        False |       False |  10.91 ms |  0.085 ms |  0.056 ms |  0.02 |    I |          - |         - |         - |    174.26 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |        False |       False |  25.68 ms |  0.181 ms |  0.120 ms |  0.04 |   II |   250.0000 |   62.5000 |         - |    8482.6 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |        False |       False | 285.16 ms |  2.967 ms |  1.963 ms |  0.41 |  III |  5000.0000 | 1000.0000 |         - | 173765.19 KB |       0.404 |
|       ParseProjectWithChangingFirstFile |        False |       False | 545.74 ms |  8.333 ms |  5.512 ms |  0.79 |   IV | 10000.0000 | 2000.0000 |         - |  341772.3 KB |       0.794 |
|          ParseProjectWithFullCacheClear |        False |       False | 690.74 ms |  7.588 ms |  4.515 ms |  1.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 430383.98 KB |       1.000 |
|                                         |              |             |           |           |           |       |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |        False |        True |  21.11 ms |  0.095 ms |  0.063 ms |  0.03 |    I |          - |         - |         - |    337.63 KB |       0.001 |
| ParseProjectWithChangingPenultimateFile |        False |        True |  38.99 ms |  0.250 ms |  0.165 ms |  0.05 |   II |   230.7692 |   76.9231 |         - |   9814.56 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |        False |        True | 346.01 ms |  4.581 ms |  2.396 ms |  0.43 |  III |  6000.0000 | 1000.0000 |         - | 199150.66 KB |       0.412 |
|       ParseProjectWithChangingFirstFile |        False |        True | 672.47 ms |  4.757 ms |  2.831 ms |  0.84 |   IV | 11000.0000 | 3000.0000 |         - |  386549.2 KB |       0.799 |
|          ParseProjectWithFullCacheClear |        False |        True | 799.92 ms |  7.737 ms |  5.117 ms |  1.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 |  483758.5 KB |       1.000 |
|                                         |              |             |           |           |           |       |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |       False |  11.15 ms |  0.063 ms |  0.037 ms |  0.02 |    I |          - |         - |         - |    176.79 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |         True |       False |  25.72 ms |  0.213 ms |  0.141 ms |  0.04 |   II |   250.0000 |   62.5000 |         - |    8480.8 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |         True |       False | 283.81 ms | 10.215 ms |  5.342 ms |  0.42 |  III |  5000.0000 | 1000.0000 |         - | 174304.71 KB |       0.406 |
|       ParseProjectWithChangingFirstFile |         True |       False | 535.58 ms |  5.894 ms |  3.508 ms |  0.79 |   IV | 10000.0000 | 2000.0000 |         - | 341390.38 KB |       0.794 |
|          ParseProjectWithFullCacheClear |         True |       False | 682.04 ms | 16.032 ms | 10.604 ms |  1.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 429729.85 KB |       1.000 |
|                                         |              |             |           |           |           |       |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |        True |  21.05 ms |  0.223 ms |  0.147 ms |  0.08 |    I |          - |         - |         - |    342.41 KB |       0.002 |
| ParseProjectWithChangingPenultimateFile |         True |        True |  31.01 ms |  1.375 ms |  0.818 ms |  0.12 |   II |   125.0000 |         - |         - |   4515.17 KB |       0.033 |
|      ParseProjectWithChangingMiddleFile |         True |        True |  84.82 ms |  0.578 ms |  0.344 ms |  0.34 |  III |   833.3333 |  333.3333 |         - |  29974.17 KB |       0.216 |
|       ParseProjectWithChangingFirstFile |         True |        True | 135.86 ms |  1.442 ms |  0.954 ms |  0.55 |   IV |  1250.0000 |  500.0000 |         - |  51778.07 KB |       0.373 |
|          ParseProjectWithFullCacheClear |         True |        True | 248.99 ms | 11.496 ms |  6.013 ms |  1.00 |    V |  1000.0000 | 1000.0000 | 1000.0000 | 138877.96 KB |       1.000 |
```

0101 commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -156,39 +157,63 @@ module IncrementalBuildSyntaxTree =
+        let semaphore = new SemaphoreSlim(1)
+        let getValue source =
+            try
+                semaphore.Wait()
```
I think the semaphore is only useful when you can use `WaitAsync` so the thread is not blocked. Here the regular lock probably does the same job.

majocha replied:
Yeah, I thought so too. This function runs fast, it doesn't really wait on anything, just fire-and-forgets the parse task, so it's not worth it to make it async. If the lock blocks, it blocks not for the entire parse duration, but just so it doesn't start the same job twice needlessly.

0101 replied:
Yeah that should be ok.

0101 commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -856,33 +856,34 @@ module IncrementalBuilderHelpers =
         let diagnosticsLogger = CompilationDiagnosticLogger("FinalizeTypeCheckTask", tcConfig.diagnosticsOptions)
         use _ = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.TypeCheck)
 
+        let! computedBoundModels = boundModels |> Seq.map (fun g -> g.GetOrComputeValue()) |> NodeCode.Parallel
```
Is running this in parallel actually helpful? Since when the bound models are created in `createBoundModelGraphNode` they are chained together sequentially.

I know that chain _can_ be broken when implementation file back by signature is edited. But it probably doesn't happen systematically.

majocha replied:
A bit but not very much. It takes the parsing part from that chain and runs it independently. Basically anytime a bound model later in that sequence would wait for a parse, it doesn't, because the results are already there.
I guess most gains would be in scenarios of full typecheck or user opening the last file in a project.

 User edits are another thing as they will cause reparsing only of the file edited.

0101 replied:
> It takes the parsing part from that chain and runs it independently

How can it do that? The first thing that `GetOrComputeValue` does is wait for previous model's `GetOrComputeValue`. How can the parsing be triggered before that finishes?

majocha replied:
Because parsing is independent from typecheck, we can do it as soon as we have the sources and we have them all at the creation of InitialState.

majocha replied:
Ah, damn it I completely missed what you were asking about and was talking about parsing, my apologies @0101.

majocha commented:
@0101, your question from yesterday got me thinking.
Don't we actually want to parallelize type check, especially in case of full type check?

If you could look at latest commit here, does it make any sense?
This should change how enablePartialTypeChecking = false situation is handled.
 My intent was to branch out full type check of sources with signatures, so that the build doesn't wait on them, but they're checked anyway.

In the benchmarks it looks like it works, the times are the same for full and partial check with signatures:

```
|                                  Method | PartialCheck | GenerateFSI |      Mean |     Error |    StdDev | Ratio | RatioSD | Rank |       Gen0 |      Gen1 |      Gen2 |    Allocated | Alloc Ratio |
|---------------------------------------- |------------- |------------ |----------:|----------:|----------:|------:|--------:|-----:|-----------:|----------:|----------:|-------------:|------------:|
|        ParseAndCheckLastFileProjectAsIs |        False |       False |  10.47 ms |  0.066 ms |  0.039 ms |  0.01 |    0.00 |    I |          - |         - |         - |    175.52 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |        False |       False |  26.25 ms |  0.459 ms |  0.304 ms |  0.04 |    0.00 |   II |   218.7500 |   62.5000 |         - |   8494.39 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |        False |       False | 293.58 ms | 17.727 ms | 11.725 ms |  0.42 |    0.02 |  III |  5000.0000 | 1000.0000 |         - | 174578.84 KB |       0.406 |
|       ParseProjectWithChangingFirstFile |        False |       False | 565.35 ms |  5.837 ms |  3.474 ms |  0.80 |    0.01 |   IV | 10000.0000 | 2000.0000 |         - |  341964.8 KB |       0.795 |
|          ParseProjectWithFullCacheClear |        False |       False | 705.47 ms | 11.687 ms |  7.731 ms |  1.00 |    0.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 430361.34 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |        False |        True |  20.64 ms |  0.198 ms |  0.131 ms |  0.08 |    0.00 |    I |          - |         - |         - |     334.4 KB |       0.002 |
| ParseProjectWithChangingPenultimateFile |        False |        True |  29.95 ms |  0.193 ms |  0.128 ms |  0.12 |    0.00 |   II |   125.0000 |   31.2500 |         - |   4526.34 KB |       0.030 |
|      ParseProjectWithChangingMiddleFile |        False |        True |  85.02 ms |  0.702 ms |  0.418 ms |  0.33 |    0.01 |  III |   500.0000 |  166.6667 |         - |  30163.14 KB |       0.202 |
|       ParseProjectWithChangingFirstFile |        False |        True | 136.68 ms |  3.286 ms |  2.174 ms |  0.53 |    0.02 |   IV |   500.0000 |  250.0000 |         - |  52518.57 KB |       0.352 |
|          ParseProjectWithFullCacheClear |        False |        True | 257.82 ms | 10.055 ms |  5.259 ms |  1.00 |    0.00 |    V |  1000.0000 | 1000.0000 | 1000.0000 | 149345.15 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |       False |  10.53 ms |  0.097 ms |  0.064 ms |  0.01 |    0.00 |    I |          - |         - |         - |     139.2 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |         True |       False |  25.71 ms |  0.170 ms |  0.112 ms |  0.04 |    0.00 |   II |   250.0000 |   62.5000 |         - |   8484.34 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |         True |       False | 294.39 ms |  9.302 ms |  6.153 ms |  0.42 |    0.01 |  III |  5000.0000 | 1000.0000 |         - | 174570.43 KB |       0.406 |
|       ParseProjectWithChangingFirstFile |         True |       False | 561.67 ms |  5.896 ms |  3.509 ms |  0.80 |    0.01 |   IV | 10000.0000 | 2000.0000 |         - | 341979.23 KB |       0.795 |
|          ParseProjectWithFullCacheClear |         True |       False | 706.11 ms | 11.092 ms |  6.601 ms |  1.00 |    0.00 |    V |  3000.0000 | 2000.0000 | 2000.0000 | 430365.53 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |            |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |        True |  20.84 ms |  0.208 ms |  0.137 ms |  0.08 |    0.00 |    I |          - |         - |         - |    333.55 KB |       0.002 |
| ParseProjectWithChangingPenultimateFile |         True |        True |  30.35 ms |  0.176 ms |  0.116 ms |  0.12 |    0.01 |   II |   125.0000 |   31.2500 |         - |   4521.43 KB |       0.032 |
|      ParseProjectWithChangingMiddleFile |         True |        True |  84.34 ms |  1.051 ms |  0.695 ms |  0.32 |    0.02 |  III |   333.3333 |  166.6667 |         - |  30142.11 KB |       0.216 |
|       ParseProjectWithChangingFirstFile |         True |        True | 137.33 ms |  2.289 ms |  1.514 ms |  0.53 |    0.03 |   IV |   250.0000 |         - |         - |  52488.88 KB |       0.376 |
|          ParseProjectWithFullCacheClear |         True |        True | 262.33 ms | 23.704 ms | 14.106 ms |  1.00 |    0.00 |    V |  1000.0000 | 1000.0000 | 1000.0000 | 139600.47 KB |       1.000 |
```

Could this potentially give us correct diagnostics for unopened implementation files, while not slowing down the IDE?

0101 commented:
@majocha If we can effectively parallelize it then we want it. Should be possible when there are signature files. Also it would be great if we could somehow use the [graph-based parallelization](https://github.com/dotnet/fsharp/pull/14494) in incremental builder. 

I'm out now, I'll have a look at your code on Monday.

majocha commented:
> If we can effectively parallelize it then we want it. Should be possible when there are signature files.

OK, I gave it another shot. 

I refactored BoundModel a bit. Removed `TcInfoState`, now there is no `PartialState `/ `FullState`, we instead create separate `GraphNode` for computing `TcInfo` and `TcInfoExtras`.
I made it more explicit when the skipping of implementation during parsing and type check happens.
When computing `TcInfo` this always skips implementations whenever possible. Computing `TcInfoExtras` OTOH always does full type check of the file, so previously skipped implementations can be checked in parallel.

This should perform better overall, especially for `enablePartialTypeChecking = false`, although I need to find a benchmark that does call `GetOrComputeTcInfoExtras()` to measure it.






> Also it would be great if we could somehow use the [graph-based parallelization](https://github.com/dotnet/fsharp/pull/14494) in incremental builder.

It would probably take some adjusting of the graph code for interactive use. If we could update the graph efficiently on source changes, it could work.

majocha commented:
What dawned on me while doing this:
It makes more sense to enclose parsing task in a GraphNode. This would give us caching for free.
(If we pretend that what we have are immutable source snapshots, which we can get away with if we start the parsing task eagerly).

T-Gro commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -104,26 +104,27 @@ type internal FSharpFile = {
 module IncrementalBuildSyntaxTree =
     open System.Runtime.CompilerServices
 
+    type ParseResult = ParsedInput * range * string * (PhasedDiagnostic * FSharpDiagnosticSeverity) array
```
Where is this type being used?

majocha replied:
In use now 🙂.

T-Gro commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -156,39 +157,58 @@ module IncrementalBuildSyntaxTree =
+        member _.Invalidate() =
+            if useCache then
+                cache.Remove(source) |> ignore
+                getValue source |> ignore
```
why do we need to call "getValue" during cache invalidation? This might start a new task, or?
(e.g. when calling Invalidate() multiple times)

majocha replied:
Yeah, repeated Invalidate would happen if the client spams with NotifyFileChanged. This can't be helped at the moment for two reasons: 
* we can't effectively cancel the parse.
* timestamp is not real versioning. Notification should come with version, like in Roslyn, (or just an `int` would do),  so we don't reparse needlessly.

majocha replied:
As for why, this is where the "eager" parsing happens.

vzarytovskii replied:
> Yeah, repeated parse would happen if the client spams with NotifyFileChanged. This can't be helped at the moment for two reasons: 
> * we can't effectively cancel the parse.
> * timestamp is not real versioning. Notification should come with version, like in Roslyn, (or just an `int` would do),  so we don't reparse needlessly.

Yeah, ideally we should be able to cancel parsing for the specific file, and have the ability to debounce requests, so we can always do it for the latest version.

T-Gro replied:
I will investigate how we can bring in cancellation.

What I would like to do even more (but I do not know how) would be common-prefix-aware debounce - i.e. cancelling and not starting from scratch, but starting from the last untouched token.

majocha replied:
I think it would be most beneficial to not parse when there are no changes. This is less of a problem with FileSystemShim, much more with live buffers in VS. VS spams a lot of workspace changed events, when a source didn't in fact change.

majocha replied:
I think this is only used in the service MRU cache, the background builder is notified just with a timestamp like this:
``` f#
            | Some builder -> do! builder.NotifyFileChanged(fileName, DateTime.UtcNow)
```
Which would be a concern if we wanted to parse eagerly and not lazily. Also, if we parse lazily, there is no guarantee of the source identity, when we do getSource.

majocha commented:
Removed eager parsing, because there's a whole bunch of tests expecting zero parses. So that's for another PR.

majocha commented:
With eager parsing removed: 
```
|                                  Method | PartialCheck | GenerateFSI |      Mean |     Error |    StdDev | Ratio | RatioSD | Rank |      Gen0 |      Gen1 |      Gen2 |    Allocated | Alloc Ratio |
|---------------------------------------- |------------- |------------ |----------:|----------:|----------:|------:|--------:|-----:|----------:|----------:|----------:|-------------:|------------:|
|        ParseAndCheckLastFileProjectAsIs |        False |       False |  10.86 ms |  0.126 ms |  0.084 ms |  0.01 |    0.00 |    I |         - |         - |         - |    145.61 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |        False |       False |  25.97 ms |  0.313 ms |  0.207 ms |  0.03 |    0.00 |   II |  250.0000 |   62.5000 |         - |   8395.83 KB |       0.020 |
|      ParseProjectWithChangingMiddleFile |        False |       False | 288.15 ms | 15.936 ms | 10.541 ms |  0.38 |    0.01 |  III | 5000.0000 | 1000.0000 |         - | 173166.74 KB |       0.402 |
|       ParseProjectWithChangingFirstFile |        False |       False | 541.13 ms |  8.423 ms |  5.571 ms |  0.72 |    0.02 |   IV | 1000.0000 |         - |         - | 338902.77 KB |       0.787 |
|          ParseProjectWithFullCacheClear |        False |       False | 754.56 ms | 17.231 ms | 11.397 ms |  1.00 |    0.00 |    V | 3000.0000 | 2000.0000 | 2000.0000 | 430527.18 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |           |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |        False |        True |  21.46 ms |  0.096 ms |  0.057 ms |  0.08 |    0.00 |    I |         - |         - |         - |    288.32 KB |       0.002 |
| ParseProjectWithChangingPenultimateFile |        False |        True |  30.59 ms |  0.262 ms |  0.173 ms |  0.12 |    0.00 |   II |  125.0000 |   31.2500 |         - |   4335.08 KB |       0.031 |
|      ParseProjectWithChangingMiddleFile |        False |        True |  71.33 ms |  1.188 ms |  0.707 ms |  0.27 |    0.00 |  III |  428.5714 |  142.8571 |         - |  28207.65 KB |       0.203 |
|       ParseProjectWithChangingFirstFile |        False |        True | 113.70 ms |  1.921 ms |  1.270 ms |  0.43 |    0.00 |   IV |  500.0000 |  250.0000 |         - |  48668.11 KB |       0.350 |
|          ParseProjectWithFullCacheClear |        False |        True | 265.33 ms |  6.493 ms |  3.396 ms |  1.00 |    0.00 |    V | 1000.0000 | 1000.0000 | 1000.0000 | 139069.17 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |           |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |       False |  10.83 ms |  0.076 ms |  0.051 ms |  0.01 |    0.00 |    I |         - |         - |         - |    153.77 KB |       0.000 |
| ParseProjectWithChangingPenultimateFile |         True |       False |  25.47 ms |  0.298 ms |  0.197 ms |  0.03 |    0.00 |   II |  218.7500 |   31.2500 |         - |   8393.53 KB |       0.019 |
|      ParseProjectWithChangingMiddleFile |         True |       False | 277.79 ms |  3.792 ms |  2.257 ms |  0.38 |    0.01 |  III | 1500.0000 | 1000.0000 |         - | 172368.18 KB |       0.400 |
|       ParseProjectWithChangingFirstFile |         True |       False | 543.20 ms |  7.595 ms |  5.023 ms |  0.74 |    0.01 |   IV | 1000.0000 |         - |         - | 339132.48 KB |       0.788 |
|          ParseProjectWithFullCacheClear |         True |       False | 736.89 ms | 10.550 ms |  6.978 ms |  1.00 |    0.00 |    V | 3000.0000 | 2000.0000 | 2000.0000 | 430567.28 KB |       1.000 |
|                                         |              |             |           |           |           |       |         |      |           |           |           |              |             |
|        ParseAndCheckLastFileProjectAsIs |         True |        True |  22.53 ms |  0.322 ms |  0.192 ms |  0.08 |    0.00 |    I |         - |         - |         - |    284.38 KB |       0.002 |
| ParseProjectWithChangingPenultimateFile |         True |        True |  30.39 ms |  0.539 ms |  0.282 ms |  0.11 |    0.00 |   II |   93.7500 |         - |         - |   4415.55 KB |       0.032 |
|      ParseProjectWithChangingMiddleFile |         True |        True |  71.22 ms |  0.854 ms |  0.565 ms |  0.26 |    0.01 |  III |  428.5714 |  142.8571 |         - |   28206.2 KB |       0.203 |
|       ParseProjectWithChangingFirstFile |         True |        True | 109.25 ms |  2.750 ms |  1.819 ms |  0.40 |    0.01 |   IV |  400.0000 |         - |         - |   48574.5 KB |       0.350 |
|          ParseProjectWithFullCacheClear |         True |        True | 275.97 ms | 12.979 ms |  6.788 ms |  1.00 |    0.00 |    V | 1000.0000 | 1000.0000 | 1000.0000 | 138795.01 KB |       1.000 |
```

Only performance with backing signatures is improved.

So eager parsing would speed up things a bit more especially when there are no .fsi files to help parallelize the build.

majocha commented:
Updated the description.

majocha commented:
/azp run

majocha commented:
ah right. There was some unrelated fail in editor integration tests.

vzarytovskii commented:
Have you, by any chance, tested how GC pressure changed on heavy codebases (like compiler itself)?

majocha commented:
> Have you, by any chance, tested how GC pressure changed on heavy codebases (like compiler itself)?
> 
> I see allocations are lower, but curious about the longevity of objects and how often gc is being kicked in

Is there a benchmark I could run for it?

nojaf commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -235,184 +226,199 @@ type TcInfoExtras =
-    member this.TryGetFull =
-        match this with
-        | TcInfoNode(_, full) -> full.TryPeekValue
+module private ValueOption =
```
Nitpick but I believe the `private` modifier is redundant if this is not listed in the signature file.

majocha commented:
@vzarytovskii, so I tried running some PerfView on this PR vs main running in VS.
The same scenario, to exercise background compiler a bit:
Open FSharp.Compiler.Service.sln
open service.fs and activity.fsi, wait for typecheck
break activity.fsi, wait for errors to propagate
fix it, wait for errors to go away.

Unfortunately I don't think I got anything representative or reproducible. In general my impression is that this PR allocates a bit more in real life (probably because it keeps the parsing results alive).

Anyway here's an example of that scenario for main:
![image](https://user-images.githubusercontent.com/1760221/231098858-5a747a62-c1c6-416d-8114-67bb4ec08e4b.png)

and this PR:
![image](https://user-images.githubusercontent.com/1760221/231098958-c3569097-e2e7-4f96-a808-3a343c557732.png)

majocha commented:
OK, the measurements in VS are worthless.
I'm randomly hitting a condition where more than one incremental builder is created for the same project and it can be observed how the same files are being processed few times in a row. That of course randomly cause a lot more allocations.
I'm not sure if this is VS specific. I will have to look into it.

majocha commented:
Making it wait for the semaphore causes the problem to be offloaded reliably to the `service.fs` cache of incremental builders:
![image](https://user-images.githubusercontent.com/1760221/231187077-0e60cf17-4ea4-4d34-a8d8-970c0f2217e3.png)

Interesting.

majocha commented:
So it looks like it was a glitch of Visual Studio loading projects after all.

majocha commented:
@vzarytovskii, I think I finally managed to capture some more representative GC data.
Repeated edits in one file when observing another in FSharp.Editor.sln, to make the background compiler churn.
So this is main:
![image](https://user-images.githubusercontent.com/1760221/231382223-fb9f0787-f126-4c07-860b-8f7211ea8a3c.png)
![image](https://user-images.githubusercontent.com/1760221/231382504-edacb56d-6a17-41ba-b4fc-730a8d22988a.png)



this PR:
![image](https://user-images.githubusercontent.com/1760221/231382349-89cbddff-e335-4007-9033-b419edbc2d95.png)
![image](https://user-images.githubusercontent.com/1760221/231382605-3204b47c-8064-49a3-9a11-b3e2861b143c.png)


While there is less activity over all, it seems a lot more survive to gen2.

vzarytovskii commented:
@majocha hm, interesting, it seems that pauses are longer in general.

majocha commented:
> @majocha hm, interesting, it seems that pauses are also longer in general.

Exactly, I'm seeing about 3x more pauses > 200ms in the stats.
Some large chunks of data being retained that are not in main? I'll try to look into it.

majocha commented:
@vzarytovskii so this is funny, actually forcing tcInfoExtras to compute
```F#
    let createBoundModelGraphNode (prevBoundModel: GraphNode<BoundModel>) syntaxTree =
        GraphNode(node {
            let! prevBoundModel = prevBoundModel.GetOrComputeValue()
            prevBoundModel.GetOrComputeTcInfoExtras() |> Async.AwaitNodeCode |> Async.Ignore |> Async.Start
            return! prevBoundModel.Next(syntaxTree)
        })
```


make the pauses go away, and the profile looks like this.
![image](https://user-images.githubusercontent.com/1760221/231396585-e78d1495-2357-4a58-94a9-3dda9a38b9b2.png)

vzarytovskii commented:
> @vzarytovskii so this is funny, actually forcing tcInfoExtras to compute
> 
> ```fsharp
>     let createBoundModelGraphNode (prevBoundModel: GraphNode<BoundModel>) syntaxTree =
>         GraphNode(node {
>             let! prevBoundModel = prevBoundModel.GetOrComputeValue()
>             prevBoundModel.GetOrComputeTcInfoExtras() |> Async.AwaitNodeCode |> Async.Ignore |> Async.Start
>             return! prevBoundModel.Next(syntaxTree)
>         })
> ```
> 
> make the pauses go away, and the profile looks like this. ![image](https://user-images.githubusercontent.com/1760221/231396585-e78d1495-2357-4a58-94a9-3dda9a38b9b2.png)

Hm, well, this is interesting.

majocha commented:
I redid some benchmarks and the gains with .fsi files and fast find references are still good. I installed this in VS and It feels fine.

TIHan reviewed (approved):
This is great work @majocha. 

Glad to see there is progress being made in `IncrementalBuilder`. It used to be one of the more difficult areas to make changes in.

0101 commented:
/azp run

0101 commented on src/Compiler/Service/IncrementalBuild.fs:
```diff
@@ -1269,6 +1132,11 @@ type IncrementalBuilder(initialState: IncrementalBuilderInitialState, state: Inc
             do! setCurrentState currentState cache ct
         }
 
+    let checkFileTimeStampsSynchronously cache =
+        checkFileTimeStamps cache
+        |> Async.AwaitNodeCode
+        |> Async.RunSynchronously
```
@majocha I started seeing some UI hangs recently in VS, specifically when finding references. Dump of the process shows that it's waiting on a task that involves this particular `RunSynchronously`. In this case checking project stamp through `IProjectReference`. 

Since this is on some paths that can be hit from different jobs quite frequently in parallel (when I was finding references semantic classification, document highlights and quick info services were also probably running), we might need to do something about it.

I saw that `IProjectReference.TryGetLogicalTimeStamp` is not async, but maybe we could just make it async? Are there any other reasons we need this to be synchronous?

majocha replied:
This is synchronous to keep type signature of call sites, only because it was all synchronous before. IIRC there was a temporary state computed instead of currentState, sometimes duplicating work.
I'll take a look. If this can't be made async we can try reverting this back to computing temp state.

majocha replied:
@0101 I think the problem is that currentState is guarded by semaphore while the tempState was not.

majocha replied:
Low cost attempt to fix this: #15146

0101 replied:
Thanks for looking into it!

majocha replied:
I think there may be something else going on here, too.
updating state in IncrementalBuilder should be instant when there are no actual changes and also super fast when there are changes. It just updates the graph, which is cold - won't do any work until requested to.

0101 replied:
There might, but avoiding `RunSynchronously` within async operations is still a good rule to keep if we can. 

I don't really get why it got stuck on this call. The dump shows UI thread waiting on this, which is being blocked by something - not sure what. And it originates from `IFSharpDocumentDiagnosticAnalyzer.AnalyzeSemanticsAsync`. 

![image](https://user-images.githubusercontent.com/217092/234811400-480e9e8a-7c8e-4e01-8d1d-54db2e85529f.png)

But I only ever saw the hangs when finding references, so there might be some weird interaction between those. 

I'm running now with your update so let's see if I see it again.

majocha replied:
I tried removing the semaphore completely, and it still blocks UI quite a lot in find references, at leas on my meager PC.

majocha replied:
I also experience this on 17.6.0 Preview 5.0, which I think does not include this PR? I kinda assumed search in VS was always causing UI lags a bit.

0101 replied:
Yeah some lags were always there, but with this my VS was completely frozen for about 10 minutes.

majocha replied:
Ah, I see, a total hang that's completely different.

0101 replied:
And I've now seen it again - with this already removed :( 

Though it was slightly different, the result was still the same, frozen VS. 

But this time it also ate all my memory. And the dump of the process shows 18 typechecks running and UI waiting for `AddSyntacticClassificationsAsync`.

![image](https://user-images.githubusercontent.com/217092/234837000-1b6d9bfb-82a6-42a8-8b24-ab4a195a8327.png)

So now I'm wondering if it could be also related to:
https://github.com/dotnet/fsharp/blob/9627d3373ee6c4c7dbb09295f5c8230cde303eea/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L109

Although that's been there for a while and I haven't seen these issues until very recently.

0101 replied:
But maybe with parsing now inside the BuildGraph it's acting up.

0101 replied:
But also now it should be quite easy to make it async - I'm going to try that.

majocha replied:
I have a very faint suspicion, because I experienced weird behaviour recently, but in CI. I'm probably worng but maybe it's the same thing? Find references is very parallel from the get go with all the Task.WhenAll and so happens that this GetHashCode() here:
https://github.com/dotnet/fsharp/blob/9627d3373ee6c4c7dbb09295f5c8230cde303eea/vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs#L331
 is very susceptible to collisions when you create many at the same time and we use it for identity. This is grasping at straws but I'd put `None` there just to try if it helps.

0101 replied:
Hmm, I'm not even sure what's the point of that. If we have 2 project options where everything is the same except stamp - why do we care which one we use?

majocha replied:
The situation I had on mind is different project options getting the same stamp, corrupting service caches. But this is very far fetched.

Is there a way to try to trigger this bug? Just spam shift-f12 in VisualFSharp.sln?

0101 replied:
Oh I see, I read the description of `Stamp` wrong. But then I would expect just flat out errors or crashes if that happened. 

I think best chances to reproduce this are shift-F12 closely before or after changing some related files, or maybe newly opening them.

safesparrow commented:
>Removed parsing cache because it's now irrelevant. Parsing step has it's own GraphNode which ensures we never reparse unless the client or the file system notifies there was source change.

@majocha Does that handle cases where project is reloaded without any changes that affect parsing results? Does it handle all types of parsing requests made by the client, either direct or indirect, or only the ones done by the incremental builder?

My suspicion is a fully working cache would still be beneficial in several cases.

majocha commented:
Exactly my thoughts, after looking into it some more. The cache in question was handling only incremental builder requests and lifetime was tied to incremental builder instances( pretty much works the same now), but incremental builder instances are efemeral by design, if I understand it correctly.
 Otoh IBs provide a built in caching through graphnodes, and it was easy to reuse. It probably would be appropriate to cache parsing on a lower level than that.

majocha commented:
Especially when considering dependency graph based type checking, it would be crucial to not reparse the whole project each time the graph needs to be obtained.
